File: OCaml/0003.ml
let largest_prime_factor (n : int64) =
  let rec helper num factor =
    if num = 1L then factor
    else if num < Int64.mul factor factor then num (* Remaining number is prime *)
    else if Int64.rem num factor = 0L then
      helper (Int64.div num factor) factor (* Keep dividing by the same factor *)
    else
      helper num (Int64.add factor 2L) (* Move to the next odd factor *)
  in
  let n = if Int64.rem n 2L = 0L then Int64.div n 2L else n in
  helper n 3L

let () =
  let number = 600851475143L in
  let result = largest_prime_factor number in
  Printf.printf "The largest prime factor of %Ld is %Ld\n" number result

File: Haskell/0003.hs
-- Function to find the largest prime factor
largestPrimeFactor :: Integer -> Integer
largestPrimeFactor n = go n 0 2
  where
    -- Remove factors of 2
    go remaining _ divisor | divisor * divisor > remaining = if remaining > 2 then remaining else 2
    go remaining largest 2 =
      if remaining `mod` 2 == 0
        then go (remaining `div` 2) 2 2
        else go remaining 2 3
    go remaining largest divisor =
      if remaining `mod` divisor == 0
        then go (remaining `div` divisor) divisor divisor
        else go remaining divisor (divisor + 2)

main :: IO ()
main = do
  let number = 600851475143
  let result = largestPrimeFactor number
  putStrLn $ "The largest prime factor of " ++ show number ++ " is " ++ show result